### ISSUE TITLE
Make JSExportedDartFunction generic and add a toJSGeneric function

### ISSUE THREAD

**Issue by srujzs** [dart-lang/sdk#54557]
It is useful to know the type of the Dart function that is now being wrapped by a JS function. Imagine an interface like:

```dart
@JS()
external set callback(JSFunction f);
```

where `callback` is called later at some point. The `external` member has no way of validating that `f` is a function that will work with the inputs that are passed and produces an output that is expected. If `f` wraps a function with the wrong type, this results in a runtime error later. In order to have some better static checking, we essentially need to duplicate the member:

```dart
@JS('callback')
external set _callback(JSFunction f);
set callback(void Function(int) f) => _callback = f.toJS;
```

Instead, we can improve the static checking here by making `JSExportedDartFunction` generic with a bound of `Function` and exposing a member that produces the generic type e.g.

```dart
JSExportedDartFunction<T> toJSGeneric<T extends Function>(T t) => t.toJS as JSExportedDartFunction<T>;
```

which would then result in code like:

```dart
@JS()
external set callback(JSExportedDartFunction<void Function(int)> f);
```

Granted, unlike with the example with two setters, this doesn't stop users from casting one `JSExportedDartFunction` to another `JSExportedDartFunction` and basically lie about the generics, but this makes it statically obvious what JS is expecting.

**Comment by srujzs** [comment]
cc @ditman because I know you came across this often with factories in Flutter.

**Comment by srujzs** [comment]
There is a downside here in that if we adopt this in `package:web`, it is a breaking change. It's also worth mentioning that if we expect a `JSExportedDartFunction` in `package:web` for all callbacks, that means non-exported functions are no longer accepted, which might not be what we want.

**Comment by ditman** [comment]
(This reminds me a little bit of the trick I used for some tests: `.jsifyAs<T extends JSAny?>()`.)

I don't have a strong preference on this one, I'm getting used to using JS*types in all `external` definitions, and when I want more ergonomic/specific Dart types to split the definition and have just a delegated call with all `.toJS`/`.toDart` called internally.

For ergonomics, this is much nicer, but the more we hide the fact that we're interfacing with JS, the more cryptic errors may become in Wasm targets, **maybe?**

**Comment by AlexDochioiu** [comment]
```dart
@JS()
external set callback(JSExportedDartFunction<void Function(int)> f);
```
+1 for this approach over the two setters approach. 

I expect the 2 setters approach could be problematic if you have something like:

```dart
@JS('removeListener')
external void _removeListener(JSFunction f);
@JS('addListener')
external void _addListener(JSFunction f);

void removeListener(void Function(int) f) => _removeListener(f.toJS);
void addListener(void Function(int) f) => _addListener(f.toJS);
```

This is what I think will happen. I did not test this so please correct me if my assumption is wrong:
Calling addListener and removeListener on the same dart function will yield a different instance of `JSExportedDartFunction`, causing the removeListener to do nothing. 

----
Having a typed JSExportedDartFunction that gets passed directly (and not converted internally through another function from Dart) may make this mistake/oversight less likely to happen. (of course, it's still possible to happen if toJS is called twice)

**Comment by ditman** [comment]
> This is what I think will happen. I did not test this so please correct me if my assumption is wrong.

That's exactly what you need today, yep.

**Comment by srujzs** [comment]
> Calling addListener and removeListener on the same dart function will yield a different instance of JSExportedDartFunction, causing the removeListener to do nothing.

Right. There's some logic on the JS backends to make sure you don't accidentally rewrap the same function, but that doesn't help here. In general, adding and removing listeners is wonky and I'd rather users try to use some of the `Stream` helpers in this package.

**Comment by ditman** [comment]
Oooo, this being closed is interesting! I haven't written a lot of js-interop in a while, but can we use this now to simplify the js-interop definitions??

**Comment by srujzs** [comment]
If you still have some of the redirecting factories for type safety, yes! It does require users calling `toJS` and `toDart` like before though, just with extra type safety.

**Comment by ditman** [comment]
We probably do, because it's likely that I haven't touched much js-interop code since this issue was created 😅
